runtime/lua/vim/lsp/_watchfiles.lua

```diff
@@ -0,0 +1,486 @@
+  -- Since Lua's built-in string pattern matching does not have an alternate
+  -- operator like '|', `parse` will construct one pattern for each possible
+  -- alternative. Any pattern that matches thus matches the glob.
+  for _, p in ipairs(pattern) do
+    if s:match('^' .. p .. '$') then
+      return true
+    end
+  end
+  return false
+end
+
+local path_sep
+
+---@private
+--- Joins filepath elements by platform-specific separator.
+---
+---@param args ... The path elements. The first element must be absolute.
+---@return string The joined path.
+local function filepath_join(...)
+  return table.concat({ ... }, path_sep)
+end
+
+local registrations = {}
+
+-- Cache of libuv handles per directory, per LSP client, per registration ID.
+local watched_paths = {}
+
+local change_queue = {}
+local change_cache = {}
+local queue_timers = {}
+
+-- Recursive flag for libuv watcher not implemented on linux
+local recursive_watch = vim.fn.has('mac') == 1 or vim.fn.has('win32') == 1
+
+-- kqueue requires one watch for every file
+local watch_each_file = vim.fn.has('bsd') == 1
+
+---@private
+--- Creates callbacks invoked on watched file events.
+---
+---@param watch_path string Absolute path being watched.
+---@param client_id number The LSP client's ID.
+---@param reg_id string The ID used to register the request.
+---@return function The callback invoked on watched file events.
+local function get_callback(watch_path, client_id, reg_id)
+  return function(path, type)
+    local matches_filter = false
+    local filters = watched_paths[watch_path].callbacks[client_id][reg_id].filters
+    for _, filter in ipairs(filters) do
+      if M._match(filter.pattern, path) and math.floor(filter.kind / (2 ^ (type - 1))) % 2 == 1 then
+        matches_filter = true
+        break
+      end
+    end
+    if not matches_filter then
+      return
+    end
+
+    local change = {
+      uri = vim.uri_from_fname(path),
+      type = type,
+    }
+    change_cache[client_id] = change_cache[client_id] or {}
+    local last_type = change_cache[client_id][change.uri]
+    if not last_type or last_type ~= change.type then
+      change_queue[client_id] = change_queue[client_id] or {}
+      table.insert(change_queue[client_id], change)
+      change_cache[client_id][change.uri] = change.type
+    end
+    if not queue_timers[client_id] then
+      queue_timers[client_id] = vim.defer_fn(function()
+        vim.lsp.get_client_by_id(client_id).notify('workspace/didChangeWatchedFiles', {
+          changes = change_queue[client_id],
+        })
+        change_queue[client_id] = nil
+        change_cache[client_id] = nil
+        queue_timers[client_id] = nil
+      end, 100)
+    end
+  end
+end
+
+-- Never create watchers for directories matching these patterns. Similar
+-- to vscode's files.watcherExclude setting.
+local excludes = {
+  parse('**/.git/objects/**'),
+  parse('**/.git/subtree-cache/**'),
+  parse('**/node_modules/*/**'),
+  parse('**/.hg/store/**'),
+}
```

Comment:
I'm not sure if we want to hardcode these like this. Could we leave this out for now, or would that affect performance too much?

Reply:
or make it overridable.

Reply:
I don't feel this get any slower testing in a medium-sized git repo, so I think it's reasonable to omit this for now.